### PR TITLE
FIX: hanging stream identification #189

### LINKED ISSUE #189: Fido hanging on skeleton stream (fmt/1000)
Bug spotted while writing docs, the following will hang when calling `identify_stream(...)`.

```python
# -*- coding: utf-8 -*-

"""Reference for Fido-based format identification

1. Create a byte-stream with a known magic number.
2. Call identify_stream(...) to identify the file against Fido's known formats.
"""

from __future__ import print_function

import io

from fido import fido

# Magic-number for fmt/1000.
BIN = b"\x5A\x58\x54\x61\x70\x65\x21\x1A\x01"

# Create the stream object with the known magic-number.
fstream = io.BytesIO(BIN)

# Create a Fido instance and call identify_stream. The identify_stream function
# will work on the stream as-is. This could be an open file handle that the
# caller is managing for itself.
f = fido.Fido()
f.identify_stream(fstream, "filename to display", extension=False)
```

I haven't looked at the code myself, but Carl has and identified we're getting to the end of the data during a loop and then it's blocking as we've nothing left to consume, but no way to exit. And it sounds like he has a fix, so that's good!

### PR DESCRIPTION
**Pull request by carlwilson**
- added simple test, with no assert, for file identification; and
- added similar for stream identification which demonstrates hang.

### CODE REVIEW

**Comment by carlwilson** [comment]
New commit added that fixes the loop and stops the tests failing.

File: tests/test_fido.py [hunk 1]
```diff
@@ -1,13 +1,50 @@
...
+    tmp_file = tmp[1]
+
+    # Write to the file our known magic-number.
+    with open(tmp_file, "wb") as new_file:
+        new_file.write(MAGIC)
+
+    # Create a Fido instance and call identify_file. The identify_file function
+    # will create and manage a file for itself.
+    f = fido.Fido()
+    f.identify_file(tmp_file)
```
Comment (sevein) [inline tests/test_fido.py:37]: Nice tests, Carl. I think this one could be a good use case for the [`tmp_path` fixture](https://docs.pytest.org/en/latest/tmpdir.html#the-tmp-path-fixture) provided by pytest. It gives us a `pathlib.Path` object which has a nice API. pytest also takes care of removing old temporary directories, always [keeping the last three](https://docs.pytest.org/en/latest/tmpdir.html#the-default-base-temporary-directory).

```suggestion
def test_file_identification(tmp_path):
    """Reference for Fido-based format identification
        1. Create a byte-stream with a known magic number and serialise to tempfile.
        2. Call identify_file(...) to identify the file against Fido's known formats.
    """
    # Create a temporary file on the host operating system.
    tmp_file = tmp_path / "file"

    # Write to the file our known magic-number.
    tmp_file.write_bytes(MAGIC)

    # Create a Fido instance and call identify_file. The identify_file function
    # will create and manage a file for itself.
    f = fido.Fido()
    f.identify_file(str(tmp_file))
```

Comment (carlwilson) [reply tests/test_fido.py:37]: So the original ideas here came from @ross-spencer (that's intended to give credit not share blame for temp file management). I agree that these could be better for test purposes. The PR will remain draft for now as I'm going to test a couple of ideas around the API.

**Comment by ross-spencer** [comment]
@carlwilson I hadn't seen @sevein's inline suggestion when I put this on my todo list. I've only just registered this coming back to it tonight but I had already created a [branch and PR](https://github.com/openpreserve/fido/pull/191) for you to take a look at here, which does the following too:

* Demos `tmp_path` like @sevein above: https://github.com/openpreserve/fido/pull/191/commits/18de5c0aa799ccd825f330a21055c930d1e4ad0b
* Brings in Pytest's [`capsys`](https://docs.pytest.org/en/latest/capture.html) to  allow us to consume the CSV and make assertions about it: https://github.com/openpreserve/fido/pull/191/commits/3b9e73f0bbd40590575176d137cd2e8e11cbb6a8

They might be good enough for you to merge, but maybe also develop out into some negative FIDO result testing too, i.e. negative identifications, if you have time/interest which it seems like you do (interest anyway, if not heaps of time! :wink: )

There was a build warning I noticed late on and corrected as well in that series of commits. 

Anyhow, I hope this work is a useful platform and not an annoyance at all. I just happened to have time at the end of my day today to take a look.

**Comment by carlwilson** [comment]
Closed as contained in #226

**Pull request closed without merging**